### ISSUE TITLE
3 Critical Issues to be fixed before one could use PgSync for production

### ISSUE THREAD

**Issue by CE-Creates** [toluaina/pgsync#62]
I would like to add the result of my findings here after playing around with pgsnc daily for almost 2 months. 
Pgsync works well when it comes to syncing some tables to elastic search. It uses replication slots to achieve this.

There are 3 major issues we are facing with pgsync now. 

1) The daemon runs smoothly, but it doesnt seem to update the elastic search. However if the pgsnc command is ran without --daemon, it picks up the changes we have done to the database. This behaviour is strange and we are currently calling the pgsync command via cron every 5 minutes as a work around .

2)  PgSync is not clearing replication slot. For us the replication slot shot up to 40GB in couple of days even when the pgsync command is run every 5 minutes on server with a cron. 

I am not an expert with PostgreSQL replication slots. But I assume that when replication slot is switch on, it creates slots for data changes in every table. In our case, we have 40+ tables out of which only 10 are used for pulling data for giving it to elastic search. We had set alarm on Cloud Watch to monitor the storage of db and we were flooded with alerts after 2 days. 
We could see that the pgsync command was running every 5 minutes with cron and the logs says it didnt find any item to index. However, the replication slot size shot up to 40GB in 2 days. **If our assumptions are correct, pgsync should clear the replication slot of other unused tables every time it polls for changes.** 

3) KeyError Issue - https://github.com/toluaina/pgsync/issues/59

@toluaina  I request you to look into this issue asap.

**Comment by toluaina** [comment]
Thanks for this useful analysis. 
I was under the impression that the key error was the only issue outstanding. 

Regarding issue:

1) From our conversation, I got the impression this worked fine in your staging environment but not in prod. I assumed this was an environmental issue. 
Are you now saying this behavior is experienced in all environments?
Can you run pgsync with verbose mode enabled and send the output?

2) I am confused as from our conversations I added REPLICATION_SLOT_CLEANUP_INTERVAL
which should cleat the replication slots. Can you confirm this is set and working as expected?
If set what value is this set to?

3) I am unable to reproduce this Key error?
Can you possibly share your DB schema and pgsync config?

**Comment by toluaina** [comment]
In addition, replication slots are not created per table but tied to the database. 
Can you list all the replication slots you have in use and the associated sizes?
 
@CE-Creates

**Comment by CE-Creates** [comment]
@toluaina  : thanks for your quick response. 

```SELECT slot_name, pg_size_pretty(pg_wal_lsn_diff(pg_current_wal_lsn(),restart_lsn)) AS replicationSlotLag, active FROM pg_replication_slots; ```

gives me the following results yesterday before I cleared the replication slot manually and reindex

```
slot_name | replicationSlotLag | status
hs_prod | 24GB | active
```

where hs is the app name and prod is the index name

**Comment by CE-Creates** [comment]
@toluaina  I have emailed you the pgsync config file and db structure dump. Hope it helps.

**Comment by toluaina** [comment]
OK, that is a start. Are you clearing the replication slots manually?
It is currently 24GB (I recall you mentioned it was 40GB at one point).
Also, have you set `REPLICATION_SLOT_CLEANUP_INTERVAL`? the default is 3600.0 and I suspect this is what is needed to clean up your replication slots.

**Comment by toluaina** [comment]
@CE-Creates I would like to propose a hangout session?

**Comment by cl0ckwork** [comment]
Chiming in here, `daemon` does not seem to work in a local environment as well, as far as I can tell an item gets added to the cache, but never gets collected, the only way to fetch pending changes is to rerun the script (via cron as suggested above)

verbose also does not seem to turn on the debug logging outside of printing some SQL

**Comment by CE-Creates** [comment]
> OK, that is a start. Are you clearing the replication slots manually?
> It is currently 24GB (I recall you mentioned it was 40GB at one point).
> Also, have you set `REPLICATION_SLOT_CLEANUP_INTERVAL`? the default is 3600.0 and I suspect this is what is needed to clean up your replication slots.

We have REPLICATION_SLOT_CLEANUP_INTERVAL set to 900.

**Comment by CE-Creates** [comment]
> @CE-Creates I would like to propose a hangout session?

That would be great. When would you be available. I am on Gulf Standard Time (GMT +3)

**Comment by CE-Creates** [comment]
> Chiming in here, `daemon` does not seem to work in a local environment as well, as far as I can tell an item gets added to the cache, but never gets collected, the only way to fetch pending changes is to rerun the script (via cron as suggested above)
> 
> verbose also does not seem to turn on the debug logging outside of printing some SQL

we haven't faced any issues locally. The daemon picked all changes on time. 
I had some trouble with the way I had written the configuration file. 
@toluaina had updated the documentation which covers pretty much everything. you might want to go through them once again.

**Comment by toluaina** [comment]
- @CE-Creates Feel free to send me an invite for any issues you need support with.
- The KeyError has been resolved in v 1.1.24.
- REPLICATION_SLOT_CLEANUP_INTERVAL of 900 implies the replication slot will be cleaned up every 15 minutes

**Comment by CE-Creates** [comment]
Thanks to @toluaina for the support.
I was able to fix all 3 problems now.

`The daemon runs smoothly, but it doesnt seem to update the elastic search. However if the pgsnc command is ran without --daemon, it picks up the changes we have done to the database. This behaviour is strange and we are currently calling the pgsync command via cron every 5 minutes as a work around .`

Running the pgsync (latest version 1.1.24) solved this issue. its now picking up changes. I could see the pgsync daemon polls picking up cache changes.

`PgSync is not clearing replication slot. For us the replication slot shot up to 40GB in couple of days even when the pgsync command is run every 5 minutes on server with a cron.`

The replication slots won't be consumed unless the pgsync is running as a daemon ( we are currently running it as nohup and would eventually do it systemd )

`KeyError Issue`

This is fixed in 1.1.24
